engine/buildkit/blob.go

```diff
@@ -5,109 +5,90 @@ import (
 // DefToBlob converts the given llb definition to a content addressed blob valid for the
 // duration of the current session. It's useful for converting unstable sources like local
 // dir imports into stable, content-defined sources.
-// NOTE: it's currently assumed that the provided definition is a single layer. Definitions
-// can be squashed into a single layer by copying from them to scratch.
 func (c *Client) DefToBlob(
 	ctx context.Context,
 	pbDef *bksolverpb.Definition,
-	compressionType compression.Type,
-) (_ *bksolverpb.Definition, desc specs.Descriptor, _ error) {
-	if compressionType == nil {
-		compressionType = compression.Zstd
-	}
-
+) (digest.Digest, error) {
 	res, err := c.Solve(ctx, bkgw.SolveRequest{
 		Definition: pbDef,
 		Evaluate:   true,
 	})
 	if err != nil {
-		return nil, desc, err
+		return "", err
 	}
 	resultProxy, err := res.SingleRef()
 	if err != nil {
-		return nil, desc, fmt.Errorf("failed to get single ref: %w", err)
+		return "", fmt.Errorf("failed to get single ref: %w", err)
 	}
 	cachedRes, err := resultProxy.Result(ctx)
 	if err != nil {
-		return nil, desc, wrapError(ctx, err, c)
+		return "", wrapError(ctx, err, c)
 	}
 	workerRef, ok := cachedRes.Sys().(*bkworker.WorkerRef)
 	if !ok {
-		return nil, desc, fmt.Errorf("invalid ref: %T", cachedRes.Sys())
+		return "", fmt.Errorf("invalid ref: %T", cachedRes.Sys())
 	}
 	ref := workerRef.ImmutableRef
 
-	// Force an unlazy of the copy in case it was lazy due to remote caching; we
-	// need it to exist locally or else blob source won't work.
-	// NOTE: in theory we could keep it lazy if we could get the descriptor handlers
-	// for the remote over to the blob source code, but the plumbing to accomplish that
-	// is tricky and ultimately only result in a marginal performance optimization.
-	err = ref.Extract(ctx, nil)
-	if err != nil {
-		return nil, desc, fmt.Errorf("failed to extract ref: %w", err)
-	}
+	dgst, _, err := checksumG.Do(ctx, ref.ID(), func(ctx context.Context) (digest.Digest, error) {
+		if err := ref.Finalize(ctx); err != nil {
+			return "", fmt.Errorf("failed to finalize ref: %w", err)
+		}
 
-	remotes, err := ref.GetRemotes(ctx, true, cacheconfig.RefConfig{
-		Compression: compression.Config{Type: compressionType},
-	}, false, nil)
-	if err != nil {
-		return nil, desc, fmt.Errorf("failed to get remotes: %w", err)
-	}
-	if len(remotes) != 1 {
-		return nil, desc, fmt.Errorf("expected 1 remote, got %d", len(remotes))
-	}
-	remote := remotes[0]
-	if len(remote.Descriptors) != 1 {
-		return nil, desc, fmt.Errorf("expected 1 descriptor, got %d", len(remote.Descriptors))
-	}
+		md := contenthash.CacheRefMetadata{RefMetadata: ref}
+		dgst, ok := md.GetContentHashKey()
+		if ok {
+			bklog.G(ctx).Debugf("DefToBlob reusing ref %s with digest %s", ref.ID(), dgst)
+			return dgst, nil
+		}
 
-	desc = remote.Descriptors[0]
+		ctx, span := Tracer(ctx).Start(ctx, fmt.Sprintf("checksum def: %s", ref.ID()), trace.WithSpanKind(trace.SpanKindInternal))
```

Comment:
I'm seeing this show up in [somewhat surprising places](https://v3.dagger.cloud/dagger/traces/6abb5fd15689b362fecd189df9109a27?span=a6f2960eec708828)

That `trace.WithSpanKind(trace.SpanKindInternal)` probably isn't doing what you want - that's some other part of OTel that we don't overload for UI purposes since all spans are actually "KindInternal" by default.

```suggestion
		ctx, span := Tracer(ctx).Start(ctx, fmt.Sprintf("checksum def: %s", ref.ID()), telemetry.Internal())
```

Reply:
Oh okay good to know, I just copy-pasta'd this from some other part of the codebase (parseRefString I think) Will fix